examples/cmd/pool.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/src-d/metadata-retrieval/github"
+)
+
+type DownloadersPool struct {
+	Size    int
+	pool    chan *DownloaderPoolItem
+	started bool
+	ended   bool
+	t0      time.Time
+	t1      time.Time
+	stats0  map[int]*downloaderStats
+	stats1  map[int]*downloaderStats
+}
+
+type DownloaderPoolItem struct {
+	id int
+	*github.Downloader
+}
+
+type DownloaderPoolStats struct {
+	Elapsed    time.Duration
+	RatesUsage map[int]*RateUsage
+}
+
+func (d *DownloaderPoolItem) stats() (*downloaderStats, error) {
+	rate, err := d.RateRemaining(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	return &downloaderStats{
+		Rate: rate,
+		Time: time.Now(),
+	}, nil
+}
+
+type downloaderStats struct {
+	Rate int
+	Time time.Time
+}
+
+type downloaderBuilder = func(c *http.Client) (*github.Downloader, error)
+
+func NewDownloadersPool(clients []*http.Client, builder downloaderBuilder) (*DownloadersPool, error) {
+	ch := make(chan *DownloaderPoolItem, len(clients))
+
+	for i, c := range clients {
+		d, err := builder(c)
+		if err != nil {
+			return nil, err
+		}
+
+		ch <- &DownloaderPoolItem{id: i, Downloader: d}
```

Comment:
why do you need to wrap downloader into `DownloaderPoolItem`?

Reply:
Because when I compute the stats for the pool I simply iterate over all the downloaders in the channel, but I need an identifier for them to be able to couple initial and ending stats.

Reply:
sorry I don't get it. Why you can't use `map[Downloader]*downloaderStats` instead?

Reply:
I guess that this shows off my python background 😂where you usually don't use objects as keys to avoid making them hashable...
so you, in this case, you use map to actually use tuples right? Seems much cleaner as it avoids using another struct 👍

Reply:
But don't you think that it will make the interface more complex? Because that woul require changing also the signature of `WithDownloader` from:
```
func (dp *DownloadersPool) WithDownloader(f func(d *DownloaderPoolItem) error) error
```
to:
```
func (dp *DownloadersPool) WithDownloader(f func(d map[*github.Downloader]*downloaderStats) error) error
```
```

Reply:
Fixed after discussion on slack, it's much simpler now! 🚀